### ISSUE TITLE
Panic when closing BatchResult

### ISSUE THREAD

**Issue by MoinTom** [jackc/pgx#1138]
In high load situations `pgx.BatchResults.Close()` sometimes causes the following panic.
```
panic: runtime error: slice bounds out of range [2322:2318]
 (empty)
 goroutine 113575 [running]:
 github.com/jackc/chunkreader/v2.(*ChunkReader).Next(0xc009c24080, 0x5)
 github.com/jackc/chunkreader/v2@v2.0.1/chunkreader.go:61 +0x365
 github.com/jackc/pgproto3/v2.(*Frontend).Receive(0xc007ee3000)
 github.com/jackc/pgproto3/v2@v2.1.1/frontend.go:82 +0x129
 github.com/jackc/pgconn.(*PgConn).peekMessage(0xc0021b6c60)
 github.com/jackc/pgconn@v1.10.0/pgconn.go:466 +0x136
 github.com/jackc/pgconn.(*PgConn).receiveMessage(0xc0021b6c60)
 github.com/jackc/pgconn@v1.10.0/pgconn.go:486 +0x25
 github.com/jackc/pgconn.(*MultiResultReader).receiveMessage(0xc0021b6d78)
 github.com/jackc/pgconn@v1.10.0/pgconn.go:1327 +0x28
 github.com/jackc/pgconn.(*MultiResultReader).Close(...)
 github.com/jackc/pgconn@v1.10.0/pgconn.go:1391
 github.com/jackc/pgx/v4.(*batchResults).Close(0xc010795400)
 github.com/jackc/pgx/v4@v4.13.0/batch.go:167 +0x1ed
 github.com/jackc/pgx/v4/pgxpool.(*poolBatchResults).Close(0xc009bc5c08)
 github.com/jackc/pgx/v4@v4.13.0/pgxpool/batch_results.go:46 +0x2d
```
It seems like that `pgproto3.Frontend.Receive()` successfully reads the next header, however the body length is 0. The following `ChunkReader.Next()` call is using a body length of `-4`.

Setup/Version:
Version: v4.13.0
Driver: pgx-pool
Target: Cockroach v21.2.3

**Comment by jackc** [comment]
I would also suggest running with the race detector.

**Comment by MoinTom** [comment]
Thanks for the tip, but i am having some issues compiling my binary with CGO enabled. So unfortunately i am not able to test  with the race detector enabled.

However i might have fixed the causing issue within my project. I found a few cases in which a defer would call `pgx.BatchResults.Close()` again. I am not having enough historical data to confirm it being fixed just yet, but it looks promising. 

I will update this issue next week when i can confirm it being fixed.

**Comment by jackc** [comment]
> However i might have fixed the causing issue within my project. I found a few cases in which a defer would call pgx.BatchResults.Close() again. I am not having enough historical data to confirm it being fixed just yet, but it looks promising.

I think that's it. Definitely wasn't safe to call `BatchResults.Close()` multiple times. Actually just bit me in production. That's a real foot-gun. I would expect that it would be safe to defer `Close()` like it is safe to defer `Rows.Close()` and `Tx.Rollback()`. I just released `v4.15.0` with a change that makes it safe.

**Comment by MoinTom** [comment]
Yes i can also confirm that it fixes the issue. Thanks for the patch, that's a welcome improvement.

Interesting enough we also saw a couple of `BUG: unexpected error from stmtCache` (`conn.SendBatch()`) errors, that disappeared after fixing the double `BatchResults.Close()`.
